src/falcon/keys.rs

```diff
@@ -0,0 +1,129 @@
+use super::{Polynomial, PublicKeyBytes, SecretKeyBytes, Signature, LOG_N};
+use crate::{hash::rpo::Rpo256, Word};
+
+#[cfg(feature = "std")]
+use super::{ffi, PK_LEN, SIG_LEN, SK_LEN};
+
+// PUBLIC KEY
+// ================================================================================================
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PublicKey(Word);
+
+impl PublicKey {
+    /// Returns a new [PublicKey] which is a commitment to the provided expanded public key.
+    ///
+    /// # Panics
+    /// Panics if the first element of the expanded public key is not 9.
+    pub fn new(pk: PublicKeyBytes) -> Self {
+        assert!(pk[0] == LOG_N as u8);
+        let h = Polynomial::from_pubkey(&pk);
+        let pk_felt = h.to_elements();
+        let pk_digest = Rpo256::hash_elements(&pk_felt).into();
+        Self(pk_digest)
+    }
+
+    // Verify
+    pub fn verify(&self, message: &[u8], signature: &Signature) -> bool {
+        signature.verify(message, self.0)
+    }
+}
+
+// SECRET KEY
+// ================================================================================================
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct SecretKey(SecretKeyBytes);
+
+impl SecretKey {
+    // Sign a message with a secret key and a seed.
+    #[cfg(feature = "std")]
+    pub fn sign(&self, message: &[u8], pk: PublicKeyBytes) -> Signature {
+        let mlen = message.len();
+        let mut sig = [0u8; SIG_LEN];
+        let mut siglen: usize = 0;
+
+        unsafe {
+            assert!(
+                ffi::PQCLEAN_FALCON512_CLEAN_crypto_sign_signature_rpo(
+                    sig.as_mut_ptr(),
+                    &mut siglen as *mut usize,
+                    message.as_ptr(),
+                    mlen,
+                    self.0.as_ptr()
+                ) == 0
+            );
+        }
+
+        Signature::new(sig, pk)
+    }
+}
+
+// KEY PAIR
+// ================================================================================================
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct KeyPair {
+    public_key: PublicKeyBytes,
+    secret_key: SecretKeyBytes,
+}
+
+impl KeyPair {
+    /// Generate a (pk, sk) keypair
+    #[cfg(feature = "std")]
+    pub fn keygen() -> Self {
```

Comment:
One of the things we'd need to support user wallets is the ability to generate keys deterministically from a given seed (i.e., `[u8; 32]`). How difficult would it be to implement this functionality?

Reply:
I have implemented it [here](https://github.com/0xPolygonMiden/crypto/blob/5a296171f117cfa59c3bcb999e644e8d7f0fbd62/src/falcon/keys.rs#L95) and changed the previous implementation on the C side to [this](https://github.com/0xPolygonMiden/crypto/blob/5a296171f117cfa59c3bcb999e644e8d7f0fbd62/src/falcon/falcon_c/falcon_rpo.c#L108).
Is `[u8; 32]` a hard requirement? The code defines a nonce for keypair generation as a `[u8; 48]`.

Reply:
If it is, then we would need to expand the short nonce to a larger one to keep with the way it is done in the reference implementation.

Reply:
I think `[u8; 48]` is fine.